### PR TITLE
add TypedCells and related functionality

### PR DESCRIPTION
**Pull request by arnej27959**
* templated DenseTensor
* templated DenseTensorModify
* add templated TypedDenseTensorBuilder
* remove DirectDenseTensorBuilder
* remove unused TensorMapper
* add dispatch structs CallMakeVector, CallDenseTensorBuilder,
  CallAppendVector, CallDotProduct, CallApply, CallReduceCells, CallJoin,
  CallSum, CallApply, CallClone, CallModify

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

This is a starting point for further discussions - not ready for merge yet.

@geirst please take a look
@toregge please take a look
@havardpe please take a look

### CODE REVIEW

File: eval/src/vespa/eval/tensor/serialization/dense_binary_format.cpp [hunk 1]
```diff
@@ -50,16 +51,16 @@ size_t decodeDimensions(nbostream & stream, std::vector<Dimension> & dimensions)
 }
 
 template<typename T, typename V>
-void decodeCells(nbostream &stream, size_t cellsSize, V &cells) {
+void decodeCells(nbostream &stream, size_t cellsSize, ArrayRef<V> cells) {
```
Comment (havardpe) [inline eval/src/vespa/eval/tensor/serialization/dense_binary_format.cpp]: consider going back to using vector to reduce changes

**Review by havardpe — COMMENTED** [review]
some random comments looking at the code

File: eval/src/vespa/eval/tensor/dense/CMakeLists.txt [hunk 2]
```diff
@@ -3,6 +3,7 @@ vespa_add_library(eval_tensor_dense OBJECT
     SOURCES
     dense_add_dimension_optimizer.cpp
     dense_dimension_combiner.cpp
+    dense_dimension_combiner.cpp
```
[the remark this answers is not in the document]
Comment (arnej27959) [reply eval/src/vespa/eval/tensor/dense/CMakeLists.txt]: fixed

File: eval/src/vespa/eval/tensor/serialization/dense_binary_format.cpp (hunk 1 above)
Comment (arnej27959) [reply eval/src/vespa/eval/tensor/serialization/dense_binary_format.cpp]: done

File: eval/src/vespa/eval/tensor/dense/typed_dense_tensor_builder.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
+
+
+#include "typed_dense_tensor_builder.h"
+
+namespace vespalib::tensor {
+
+size_t num_typed_tensor_builder_inserts = 0;
```
[the remark this answers is not in the document]
Comment (arnej27959) [reply eval/src/vespa/eval/tensor/dense/typed_dense_tensor_builder.cpp]: leftover cruft, removed

File: eval/src/vespa/eval/tensor/dense/dense_tensor_view.cpp [hunk 4]
```diff
@@ -146,27 +174,43 @@ DenseTensorView::type() const
...
-        result += cell;
-    }
-    return result;
+    return dispatch_1<CallSum>(_cellsRef);
 }
 
+struct CallApply {
+    template <typename CT>
+    static Tensor::UP
+    call(const ConstArrayRef<CT> &oldCells, eval::ValueType newType, const CellFunction &func)
```
[the remark this answers is not in the document]
Comment (arnej27959) [reply eval/src/vespa/eval/tensor/dense/dense_tensor_view.cpp]: done

File: eval/src/vespa/eval/tensor/dense/dense_tensor.cpp [hunk 5]
```diff
@@ -21,68 +21,57 @@ calcCellsSize(const eval::ValueType &type)
...
-{
-    initCellsRef(CellsRef(_cells));
-    checkCellsSize(*this);
-}
-
-DenseTensor::DenseTensor(eval::ValueType &&type_in,
-                         Cells &&cells_in)
-    : DenseTensorView(_type),
+template <typename CT>
+DenseTensor<CT>::DenseTensor(eval::ValueType &&type_in,
```
[the remark this answers is not in the document]
Comment (arnej27959) [reply eval/src/vespa/eval/tensor/dense/dense_tensor.cpp]: done

File: eval/src/vespa/eval/tensor/dense/dense_dot_product_function.cpp [hunk 6]
```diff
@@ -19,17 +18,61 @@ using namespace eval::operation;
...
 void my_dot_product_op(eval::InterpretedFunction::State &state, uint64_t param) {
     auto *hw_accelerator = (hwaccelrated::IAccelrated *)(param);
-    DenseTensorView::CellsRef lhsCells = getCellsRef(state.peek(1));
-    DenseTensorView::CellsRef rhsCells = getCellsRef(state.peek(0));
-    size_t numCells = std::min(lhsCells.size(), rhsCells.size());
-    double result = hw_accelerator->dotProduct(lhsCells.cbegin(), rhsCells.cbegin(), numCells);
+    TypedCells lhsCells = getCellsRef(state.peek(1));
+    TypedCells rhsCells = getCellsRef(state.peek(0));
+    size_t numCells = std::min(lhsCells.size, rhsCells.size);
+    double result = dispatch_2<CallDotProduct>(lhsCells, rhsCells, numCells, hw_accelerator);
```
[the remark this answers is not in the document]
Comment (arnej27959) [reply eval/src/vespa/eval/tensor/dense/dense_dot_product_function.cpp]: assuming double-only now.

**Pull request closed without merging**